File: tools/cvqr.py
"""
cvqr.py — reference implementation of the CVQR/1 container.

CVQR/1 is a self-describing binary capsule carrying Codec2-compressed mono
speech, wrapped in Base45 (RFC 9285) and prefixed with the ASCII tag `CVQR1:`
so it can live inside an ordinary QR code in alphanumeric mode.

This module has NO third-party dependencies. It is deliberately written to be
readable as a specification in its own right: see FORMAT.md for the normative
description.
"""

from __future__ import annotations

import struct
import zlib
from dataclasses import dataclass

MAGIC = b"CVQR"
VERSION = 1
PREFIX = "CVQR1:"
HEADER_LEN = 24

# ---------------------------------------------------------------------------
# Codec2 mode registry
# ---------------------------------------------------------------------------
# id -> (name, bits_per_frame, bytes_per_frame_as_written_by_c2enc, frame_ms)
#
# `bytes_per_frame` is how the reference `c2enc` tool byte-aligns each frame on
# disk. CVQR/1 does NOT store that padding: frames are packed contiguously,
# MSB-first (see pack_frames/unpack_frames). At 1300 bit/s this saves ~7% of
# the payload, which is worth a whole QR version.

MODES = {
    0x01: ("700C", 28, 4, 40),
    0x02: ("1200", 48, 6, 40),
    0x03: ("1300", 52, 7, 40),
    0x04: ("1400", 56, 7, 40),
    0x05: ("1600", 64, 8, 40),
    0x06: ("2400", 48, 6, 20),
    0x07: ("3200", 64, 8, 20),
}
MODES_BY_NAME = {v[0]: k for k, v in MODES.items()}

CODEC_CODEC2 = 0x01


class CVQRError(ValueError):
    """Raised for any malformed or unsupported CVQR payload.

    The message is intended to be shown directly to a human.
    """


# ---------------------------------------------------------------------------
# Base45 (RFC 9285)
# ---------------------------------------------------------------------------

B45_ALPHABET = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ $%*+-./:"
B45_VALUES = {c: i for i, c in enumerate(B45_ALPHABET)}


def b45_encode(data: bytes) -> str:
    out = []
    for i in range(0, len(data) - 1, 2):
        n = (data[i] << 8) | data[i + 1]
        n, c = divmod(n, 45)
        e, d = divmod(n, 45)
        out.append(B45_ALPHABET[c] + B45_ALPHABET[d] + B45_ALPHABET[e])
    if len(data) % 2:
        d, c = divmod(data[-1], 45)
        out.append(B45_ALPHABET[c] + B45_ALPHABET[d])
    return "".join(out)


def b45_decode(text: str) -> bytes:
    try:
        vals = [B45_VALUES[c] for c in text]
    except KeyError as exc:
        raise CVQRError(f"Not valid Base45: character {exc.args[0]!r} is not in the alphabet.") from None
    if len(vals) % 3 == 1:
        raise CVQRError("Not valid Base45: length leaves a single trailing character.")
    out = bytearray()
    for i in range(0, len(vals) - 2, 3):
        n = vals[i] + vals[i + 1] * 45 + vals[i + 2] * 45 * 45
        if n > 0xFFFF:
            raise CVQRError("Not valid Base45: a 3-character group exceeds 16 bits.")
        out += struct.pack(">H", n)
    if len(vals) % 3 == 2:
        n = vals[-2] + vals[-1] * 45
        if n > 0xFF:
            raise CVQRError("Not valid Base45: the final 2-character group exceeds 8 bits.")
        out.append(n)
    return bytes(out)


# ---------------------------------------------------------------------------
# Frame bit-packing
# ---------------------------------------------------------------------------

def pack_frames(raw: bytes, bits_per_frame: int, bytes_per_frame: int) -> bytes:
    """c2enc's byte-aligned frames -> contiguous MSB-first bitstream."""
    if len(raw) % bytes_per_frame:
        raise CVQRError(
            f"Codec2 file is {len(raw)} bytes, not a whole number of "
            f"{bytes_per_frame}-byte frames."
        )
    acc = 0
    nbits = 0
    out = bytearray()
    for off in range(0, len(raw), bytes_per_frame):
        frame = int.from_bytes(raw[off:off + bytes_per_frame], "big")
        # Frame occupies the TOP bits_per_frame bits of the aligned group.
        frame >>= (bytes_per_frame * 8 - bits_per_frame)
        acc = (acc << bits_per_frame) | frame
        nbits += bits_per_frame
        while nbits >= 8:
            nbits -= 8
            out.append((acc >> nbits) & 0xFF)
            acc &= (1 << nbits) - 1
    if nbits:
        out.append((acc << (8 - nbits)) & 0xFF)
    return bytes(out)


def unpack_frames(packed: bytes, n_frames: int, bits_per_frame: int, bytes_per_frame: int) -> bytes:
    """Contiguous MSB-first bitstream -> c2dec's byte-aligned frames."""
    need_bits = n_frames * bits_per_frame
    if len(packed) * 8 < need_bits:
        raise CVQRError(
            f"Payload holds {len(packed) * 8} bits but {n_frames} frames need {need_bits}."
        )
    total = int.from_bytes(packed, "big")
    total >>= (len(packed) * 8 - need_bits)  # drop trailing pad bits
    out = bytearray()
    for i in range(n_frames - 1, -1, -1):
        frame = (total >> (i * bits_per_frame)) & ((1 << bits_per_frame) - 1)
        frame <<= (bytes_per_frame * 8 - bits_per_frame)
        out += frame.to_bytes(bytes_per_frame, "big")
    return bytes(out)


# ---------------------------------------------------------------------------
# Capsule
# ---------------------------------------------------------------------------

@dataclass
class Capsule:
    mode_id: int
    duration_ms: int
    frame_count: int
    payload: bytes          # contiguously bit-packed Codec2 frames
    codec_id: int = CODEC_CODEC2
    flags: int = 0
    sample_rate: int = 8000
    channels: int = 1
    bits_per_sample: int = 16

    @property
    def mode_name(self) -> str:
        return MODES[self.mode_id][0]

    def serialize(self) -> bytes:
        if self.mode_id not in MODES:
            raise CVQRError(f"Unknown Codec2 mode id 0x{self.mode_id:02X}.")
        for name, val, limit in (
            ("duration_ms", self.duration_ms, 0xFFFFFFFF),
            ("frame_count", self.frame_count, 0xFFFF),
            ("payload length", len(self.payload), 0xFFFF),
            ("sample_rate", self.sample_rate, 0xFFFF),
        ):
            if not 0 <= val <= limit:
                raise CVQRError(f"{name} = {val} does not fit its field.")
        head = struct.pack(
            ">4sBBBBHBBIHHI",
            MAGIC, VERSION, self.codec_id, self.mode_id, self.flags,
            self.sample_rate, self.channels, self.bits_per_sample,
            self.duration_ms, self.frame_count, len(self.payload), 0,
        )
        assert len(head) == HEADER_LEN, len(head)
        blob = head + self.payload
        crc = zlib.crc32(blob) & 0xFFFFFFFF
        return blob[:20] + struct.pack(">I", crc) + blob[24:]

    @classmethod
    def parse(cls, blob: bytes) -> "Capsule":
        if len(blob) < HEADER_LEN:
            raise CVQRError(f"Truncated capsule: {len(blob)} bytes, need at least {HEADER_LEN}.")
        (magic, version, codec_id, mode_id, flags, sample_rate,
         channels, bits_per_sample, duration_ms, frame_count,
         payload_len, crc) = struct.unpack(">4sBBBBHBBIHHI", blob[:HEADER_LEN])

        if magic != MAGIC:
            raise CVQRError(f"Bad magic {magic!r}: this is not a CVQR capsule.")
        if version != VERSION:
            raise CVQRError(f"Capsule declares version {version}; this decoder implements 1.")
        if codec_id != CODEC_CODEC2:
            raise CVQRError(f"Unsupported codec id 0x{codec_id:02X}; only Codec2 (0x01) is defined.")
        if mode_id not in MODES:
            raise CVQRError(f"Unsupported Codec2 mode id 0x{mode_id:02X}.")

        payload = blob[HEADER_LEN:]
        if len(payload) != payload_len:
            raise CVQRError(
                f"Header declares a {payload_len}-byte payload but {len(payload)} bytes follow."
            )

        zeroed = blob[:20] + b"\x00\x00\x00\x00" + blob[24:]
        actual = zlib.crc32(zeroed) & 0xFFFFFFFF
        if actual != crc:
            raise CVQRError(
                f"Integrity check failed: CRC-32 is {actual:08X}, capsule claims {crc:08X}. "
                "The payload is damaged."
            )

        bits_per_frame = MODES[mode_id][1]
        need = -(-frame_count * bits_per_frame // 8)
        if payload_len != need:
            raise CVQRError(
                f"{frame_count} frames at {bits_per_frame} bits need {need} bytes, "
                f"but the payload is {payload_len}."
            )
        return cls(mode_id=mode_id, duration_ms=duration_ms, frame_count=frame_count,
                   payload=payload, codec_id=codec_id, flags=flags,
                   sample_rate=sample_rate, channels=channels,
                   bits_per_sample=bits_per_sample)


# ---------------------------------------------------------------------------
# Text wrapping
# ---------------------------------------------------------------------------

def to_text(capsule: Capsule) -> str:
    return PREFIX + b45_encode(capsule.serialize())


def from_text(text: str) -> Capsule:
    """Parse `CVQR1:`-prefixed text.

    SPACE IS A VALID BASE45 CHARACTER, so whitespace cannot simply be stripped
    — doing so silently corrupts most payloads. Only line breaks and tabs are
    removed (they can never appear in the alphabet, and do appear when a
    payload is wrapped). Edge spaces are ambiguous, so the string is tried
    as-is first and only trimmed as a fallback: real data always wins.
    """
    cleaned = text.replace("\r", "").replace("\n", "").replace("\t", "")

    def attempt(s: str) -> Capsule:
        if not s.upper().startswith(PREFIX):
            head = s[:16] + ("..." if len(s) > 16 else "")
            raise CVQRError(f"Missing the CVQR1: prefix. This payload starts {head!r}.")
        return Capsule.parse(b45_decode(s[len(PREFIX):]))

    try:
        return attempt(cleaned)
    except CVQRError:
        trimmed = cleaned.strip()
        if trimmed == cleaned:
            raise
        return attempt(trimmed)

File: decoder/cvqr.js
/*
 * cvqr.js — CVQR/1 reader. Mirrors tools/cvqr.py exactly; FORMAT.md is normative.
 * No dependencies, no network, no state.
 */
'use strict';

const CVQR = (() => {
  const MAGIC = [0x43, 0x56, 0x51, 0x52];      // "CVQR"
  const VERSION = 1;
  const PREFIX = 'CVQR1:';
  const HEADER_LEN = 24;

  // id -> [name, bitsPerFrame, bytesPerFrameAligned, frameMs, codec2ModeId]
  const MODES = {
    0x01: ['700C', 28, 4, 40, 8],
    0x02: ['1200', 48, 6, 40, 5],
    0x03: ['1300', 52, 7, 40, 4],
    0x04: ['1400', 56, 7, 40, 3],
    0x05: ['1600', 64, 8, 40, 2],
    0x06: ['2400', 48, 6, 20, 1],
    0x07: ['3200', 64, 8, 20, 0],
  };

  class CVQRError extends Error {}

  const B45 = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ $%*+-./:';
  const B45_VAL = new Map([...B45].map((c, i) => [c, i]));

  function b45Decode(text) {
    const vals = [];
    for (const ch of text) {
      const v = B45_VAL.get(ch);
      if (v === undefined) {
        throw new CVQRError(`Not valid Base45: the character "${ch}" is not in the alphabet.`);
      }
      vals.push(v);
    }
    if (vals.length % 3 === 1) {
      throw new CVQRError('Not valid Base45: the length leaves a single trailing character.');
    }
    const out = [];
    let i = 0;
    for (; i + 2 < vals.length; i += 3) {
      const n = vals[i] + vals[i + 1] * 45 + vals[i + 2] * 2025;
      if (n > 0xffff) throw new CVQRError('Not valid Base45: a 3-character group exceeds 16 bits.');
      out.push((n >> 8) & 0xff, n & 0xff);
    }
    if (vals.length % 3 === 2) {
      const n = vals[vals.length - 2] + vals[vals.length - 1] * 45;
      if (n > 0xff) throw new CVQRError('Not valid Base45: the final group exceeds 8 bits.');
      out.push(n);
    }
    return new Uint8Array(out);
  }

  // CRC-32/ISO-HDLC, the zlib/PNG/gzip one.
  const CRC_TABLE = (() => {
    const t = new Uint32Array(256);
    for (let n = 0; n < 256; n++) {
      let c = n;
      for (let k = 0; k < 8; k++) c = c & 1 ? 0xedb88320 ^ (c >>> 1) : c >>> 1;
      t[n] = c >>> 0;
    }
    return t;
  })();

  function crc32(bytes, skipFrom, skipTo) {
    let c = 0xffffffff;
    for (let i = 0; i < bytes.length; i++) {
      // bytes in [skipFrom, skipTo) are covered as zero — see FORMAT.md §5
      const b = i >= skipFrom && i < skipTo ? 0 : bytes[i];
      c = CRC_TABLE[(c ^ b) & 0xff] ^ (c >>> 8);
    }
    return (c ^ 0xffffffff) >>> 0;
  }

  function parseCapsule(blob) {
    if (blob.length < HEADER_LEN) {
      throw new CVQRError(`Truncated capsule: ${blob.length} bytes, need at least ${HEADER_LEN}.`);
    }
    const dv = new DataView(blob.buffer, blob.byteOffset, blob.byteLength);
    for (let i = 0; i < 4; i++) {
      if (blob[i] !== MAGIC[i]) {
        throw new CVQRError('Bad magic: this is not a CVQR capsule.');
      }
    }
    const version = blob[4], codecId = blob[5], modeId = blob[6], flags = blob[7];
    if (version !== VERSION) {
      throw new CVQRError(`Capsule declares version ${version}; this decoder implements 1.`);
    }
    if (codecId !== 0x01) {
      throw new CVQRError(`Unsupported codec id 0x${codecId.toString(16)}; only Codec2 (0x01) is defined.`);
    }
    if (!(modeId in MODES)) {
      throw new CVQRError(`Unsupported Codec2 mode id 0x${modeId.toString(16).padStart(2, '0')}.`);
    }
    const sampleRate = dv.getUint16(8);
    const channels = blob[10];
    const bitsPerSample = blob[11];
    const durationMs = dv.getUint32(12);
    const frameCount = dv.getUint16(16);
    const payloadLen = dv.getUint16(18);
    const crcClaimed = dv.getUint32(20);

    const payload = blob.subarray(HEADER_LEN);
    if (payload.length !== payloadLen) {
      throw new CVQRError(
        `Header declares a ${payloadLen}-byte payload but ${payload.length} bytes follow.`);
    }
    const crcActual = crc32(blob, 20, 24);
    if (crcActual !== crcClaimed) {
      throw new CVQRError(
        `Integrity check failed: CRC-32 is ${crcActual.toString(16).toUpperCase().padStart(8, '0')}, ` +
        `capsule claims ${crcClaimed.toString(16).toUpperCase().padStart(8, '0')}. The payload is damaged.`);
    }
    const [name, bits] = MODES[modeId];
    const need = Math.ceil((frameCount * bits) / 8);
    if (payloadLen !== need) {
      throw new CVQRError(
        `${frameCount} frames at ${bits} bits need ${need} bytes, but the payload is ${payloadLen}.`);
    }
    return { modeId, modeName: name, codecId, flags, sampleRate, channels,
             bitsPerSample, durationMs, frameCount, payload, crc: crcClaimed };
  }

  /* Contiguous MSB-first bitstream -> c2dec-compatible byte-aligned frames. */
  function unpackFrames(cap) {
    const [, bits, aligned] = MODES[cap.modeId];
    const out = new Uint8Array(cap.frameCount * aligned);
    let bitPos = 0;
    for (let f = 0; f < cap.frameCount; f++) {
      for (let b = 0; b < bits; b++, bitPos++) {
        const bit = (cap.payload[bitPos >> 3] >> (7 - (bitPos & 7))) & 1;
        if (bit) {
          const dst = f * aligned + (b >> 3);
          out[dst] |= 1 << (7 - (b & 7));
        }
      }
    }
    return out;
  }

  /*
   * SPACE IS A VALID BASE45 CHARACTER. Stripping whitespace to be tidy about
   * pasted text silently corrupts any payload that contains one — which is
   * most of them. So only line breaks and tabs are removed here, because
   * those cannot appear in the alphabet and do appear when a payload is
   * wrapped across lines.
   *
   * Edge spaces are genuinely ambiguous: they may be data, or they may be a
   * sloppy selection. Try the string as-is first, and only fall back to
   * trimming them if that fails — so real data always wins.
   */
  function fromText(text) {
    const cleaned = String(text).replace(/[\r\n\t]+/g, '');
    const attempt = (s) => {
      if (!s.toUpperCase().startsWith(PREFIX)) {
        const head = s.slice(0, 16) + (s.length > 16 ? '…' : '');
        throw new CVQRError(`Missing the CVQR1: prefix. This payload starts "${head}".`);
      }
      return parseCapsule(b45Decode(s.slice(PREFIX.length)));
    };
    try {
      return attempt(cleaned);
    } catch (first) {
      const trimmed = cleaned.trim();
      if (trimmed === cleaned) throw first;
      try { return attempt(trimmed); } catch { throw first; }
    }
  }

  return { MODES, PREFIX, CVQRError, b45Decode, crc32, parseCapsule, unpackFrames, fromText };
})();

if (typeof module !== 'undefined') module.exports = CVQR;
